File: mjlib_lua/base_table/hulib.lua
local table_mgr = require "table_mgr"
local M = {}

function M.get_hu_info(hand_cards, cur_card)
    local cards = {}
    for i,v in ipairs(hand_cards) do
        cards[i] = v
    end

    if cur_card then
        cards[cur_card] = cards[cur_card] + 1
    end

    local first_info = {
        eye = false,            -- 当前是否有将
    }

    -- 字牌（东西南北中发白)
    if not M.check_color(cards, first_info) then
        return false
    -- 万
    elseif not M.check_color_chi(cards, 1, 9, first_info) then
        return false
    -- 筒
    elseif not M.check_color_chi(cards, 10, 18, first_info) then
        return false
    -- 条
    elseif not M.check_color_chi(cards, 19, 27, first_info) then
        return false
    end
    return true
end

function M.check_color(cards, min, max, info)
    for i = 28, 34 do
        local count = cards[i]

        if count == 1 or count == 4 then
            return false
        end

        if count == 2 then
            if info.eye then
                return false
            end

            info.eye = true
        end
    end

    return true
end

function M.check_color_chi(cards, min, max, info)
    local key = 0
    for i = min, max do
        local c = cards[i]
        if c > 0 then
            key = key * 10 + c
		end

        if key > 0 and (c == 0 or i == max) then
            local eye = (key%3 == 2)
            if info.eye and eye then
                return false
            end
			
            if eye then
                info.eye = true
            end

            if not table_mgr:check(key) then
                return false
            end
            key = 0
        end
    end

    return true
end

return M

File: lesson/mj_gui_table/hulib.cpp
#include "stdafx.h"

#include <string.h>
#include "hulib.h"
#include "table_mgr.h"

bool HuLib::get_hu_info(char* cards, int gui_index1,int gui_index2)
{
	int gui_num1 = 0;
	int gui_num2 = 0;
	if (gui_index1 != 34) gui_num1 = cards[gui_index1];
	cards[gui_index1] = 0;

	if (gui_index2 != 34) gui_num2 = cards[gui_index2];
	cards[gui_index2] = 0;

	bool hu = split(cards, gui_num1+ gui_num2);

	if (gui_index1 != 34) cards[gui_index1] = gui_num1;
	if (gui_index2 != 34) cards[gui_index2] = gui_num2;
	return hu;
}

bool check(int gui, int eye_num, int gui_num, int& gui_sum) {
	if (gui < 0) return false;

	gui_sum += gui;
	if (gui_sum > gui_num) return false;

	if (eye_num == 0) return true;

	return gui_sum + (eye_num - 1) <= gui_num;
}

bool HuLib::split(char* const cards, int gui_num)
{
	int eye_num = 0;
	int gui_sum = 0;
	int gui = 0;
	// 检查万
	gui = _split(cards, gui_num, 0, 8, true, eye_num);
	if (!check(gui, eye_num, gui_num, gui_sum)) return false;
	// 检查筒
	gui = _split(cards, gui_num-gui_sum, 9, 17, true, eye_num);
	if (!check(gui, eye_num, gui_num, gui_sum)) return false;
	// 检查条
	gui = _split(cards, gui_num-gui_sum, 18, 26, true, eye_num);
	if (!check(gui, eye_num, gui_num, gui_sum)) return false;

	gui = _split(cards, gui_num-gui_sum, 27, 33, false, eye_num);
	if (!check(gui, eye_num, gui_num, gui_sum)) return false;

	if (eye_num == 0)
	{
		return gui_sum + 2 <= gui_num;
	}

	return true;
}

int HuLib::_split(char* const cards, int gui_num, int min, int max, bool chi, int& eye_num)
{
    int key = 0;
    int num = 0;

    for(int i = min; i <= max; ++i)
    {
        key = key*10 + cards[i];
        num = num + cards[i];
    }

	if (num == 0) return 0;

	for (int i = 0; i <= gui_num; ++i)
	{
		int yu = (num + i) % 3;
		if (yu == 1) continue;
		bool eye = (yu == 2);
		if (TableMgr::check(key, i, eye, chi))
		{
			if (eye) eye_num++;
			return i;
		}
	}

    return -1;
}

